File: Example/QDNavigationBar/QDConfigEditor/QDConfigEditorView.m
//
//  QDConfigEditorView.m
//  QDCustomNavigation_Example
//
//  Created by Sinno on 2020/8/5.
//  Copyright © 2020 sinno93. All rights reserved.
//

#import "QDConfigEditorView.h"
#import "QDConfigColorView.h"
#import "QDConfigSwitchView.h"
#import "QDConfigSegmentView.h"

#import <Masonry/Masonry.h>
#import <QDNavigationBar_Example-Swift.h>

@interface QDConfigEditorView ()
@property (nonatomic, strong) UIStackView *stackView;
@property (nonatomic, strong) QDConfigColorView *bgColorView;
@property (nonatomic, strong) QDConfigColorView *bottomLineColorView;
@property (nonatomic, strong) QDConfigSwitchView *barHiddenView;
@property (nonatomic, strong) QDConfigSwitchView *translucentView;
@property (nonatomic, strong) QDConfigSwitchView *statusBarHiddenView;
@property (nonatomic, strong) QDConfigSegmentView *statusBarStyleView;
@end

@implementation QDConfigEditorView
#pragma mark - Initialization
// 初始化方法
- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self configSubviews];
    }
    return self;
}

#pragma mark - Super methods

#pragma mark - Public Methods
// 公开的方法
- (void)setConfig:(QDNavigationBarConfig *)config {
    _config = config;
    [self updateView];
}


#pragma mark - Private Methods
// 私有的方法
- (void)configSubviews {
    [self addSubview:self.stackView];
    [self.stackView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self).insets(UIEdgeInsetsMake(8, 0, 5, 0));
    }];
    NSArray *array = @[self.bgColorView,
                       self.bottomLineColorView,
                       self.barHiddenView,
                       self.translucentView,
                       self.statusBarHiddenView,
                       self.statusBarStyleView,
                    ];
    for (UIView *view in array) {
        [self.stackView addArrangedSubview:view];
        [view mas_makeConstraints:^(MASConstraintMaker *make) {
            make.height.equalTo(@48);
        }];
    }
    self.layer.borderColor = UIColor.lightGrayColor.CGColor;
    self.layer.borderWidth = 1.0/UIScreen.mainScreen.scale;
    self.layer.cornerRadius = 15;
}

- (void)updateView {
    self.bgColorView.color = self.config.backgroundColor;
    self.bottomLineColorView.color = self.config.bottomLineColor;
    self.barHiddenView.on = self.config.barHidden;
    self.translucentView.on = self.config.needBlurEffect;
    self.statusBarHiddenView.on = self.config.statusBarHidden;
    self.statusBarStyleView.selectedIndex = ({
        NSInteger index = 0;
        switch (self.config.statusBarStyle) {
            case UIStatusBarStyleDefault:
                index = 0;
                break;
            case UIStatusBarStyleLightContent:
                index = 1;
                break;
            case UIStatusBarStyleDarkContent:
                index = 2;
                break;
            default:
                break;
        }
        index;
    });
}

#pragma mark - Getter && Setter
// Getter和Setter方法
- (UIStackView *)stackView {
    if (!_stackView) {
        UIStackView *view = [[UIStackView alloc] init];
        view.axis = UILayoutConstraintAxisVertical;
        view.alignment = UIStackViewAlignmentFill;
        view.spacing = 1.0f;
        _stackView = view;
    }
    return _stackView;
}
- (QDConfigColorView *)bgColorView {
    if (!_bgColorView) {
        QDConfigColorView *view = [[QDConfigColorView alloc] init];
        view.title = @"backgroundColor";
        view.color = UIColor.whiteColor;
        __weak __typeof(self)weakSelf = self;
        view.colorChanged = ^(UIColor * _Nonnull color) {
            weakSelf.config.backgroundColor = color;
        };
        _bgColorView = view;
    }
    return _bgColorView;
}

- (QDConfigColorView *)bottomLineColorView {
    if (!_bottomLineColorView) {
        QDConfigColorView *view = [[QDConfigColorView alloc] init];
        view.title = @"bottomLineColor";
        view.color = UIColor.redColor;
        __weak __typeof(self)weakSelf = self;
        view.colorChanged = ^(UIColor * _Nonnull color) {
            weakSelf.config.bottomLineColor = color;
        };
        _bottomLineColorView = view;
    }
    return _bottomLineColorView;
}


- (QDConfigSwitchView *)barHiddenView {
    if (!_barHiddenView) {
        QDConfigSwitchView *view = [[QDConfigSwitchView alloc] init];
        view.title = @"barHidden";
        view.on = YES;
        __weak __typeof(self)weakSelf = self;
        view.valueChanged = ^(BOOL on) {
            weakSelf.config.barHidden = on;
        };
        _barHiddenView = view;
    }
    return _barHiddenView;
}

- (QDConfigSwitchView *)translucentView {
    if (!_translucentView) {
        QDConfigSwitchView *view = [[QDConfigSwitchView alloc] init];
        view.title = @"needBlurEffect";
        view.on = YES;
        __weak __typeof(self)weakSelf = self;
        view.valueChanged = ^(BOOL on) {
            weakSelf.config.needBlurEffect = on;
        };
        _translucentView = view;
    }
    return _translucentView;
}

- (QDConfigSwitchView *)statusBarHiddenView {
    if (!_statusBarHiddenView) {
        QDConfigSwitchView *view = [[QDConfigSwitchView alloc] init];
        view.title = @"statusBarHidden";
        view.on = YES;
        __weak __typeof(self)weakSelf = self;
        view.valueChanged = ^(BOOL on) {
            weakSelf.config.statusBarHidden = on;
        };
        _statusBarHiddenView = view;
    }
    return _statusBarHiddenView;
}

- (QDConfigSegmentView *)statusBarStyleView {
    if (!_statusBarStyleView) {
        NSArray *items = nil;
        if (@available(iOS 13.0, *)) {
            items = @[@"Default", @"Light", @"Dark"];
        } else {
            items = @[@"Default", @"Light"];
        }
        QDConfigSegmentView *view = [[QDConfigSegmentView alloc] initWithItems: items];
        view.title = @"statusBarStyle";
        view.selectedIndex = 0;
        __weak __typeof(self)weakSelf = self;
        view.valueChanged = ^(NSInteger selectedIndex) {
            UIStatusBarStyle style = UIStatusBarStyleDefault;
            switch (selectedIndex) {
                case 0:
                    style = UIStatusBarStyleDefault;
                    break;
                case 1:
                    style = UIStatusBarStyleLightContent;
                    break;
                case 2:
                    if (@available(iOS 13.0, *)) {
                        style = UIStatusBarStyleDarkContent;
                    }
                    break;
                default:
                    break;
            }
            weakSelf.config.statusBarStyle = style;
        };
        _statusBarStyleView = view;
    }
    return _statusBarStyleView;
}

@end

File: Example/QDNavigationBar/QDConfigEditor/QDConfigEditorView.swift
//
//  QDConfigEditorView.swift
//  QDNavigationBar_Example
//
//  Created by zqd on 2024/6/5.
//  Copyright © 2024 CocoaPods. All rights reserved.
//

import UIKit
import SnapKit
import QDNavigationBar

class QDConfigEditorView: UIView {
    
    private var stackView: UIStackView = {
        let view = UIStackView()
        view.axis = .vertical
        view.alignment = .fill
        view.spacing = 1.0
        return view
    }()
    
    private lazy var bgColorView: QDConfigColorView = {
        let view = QDConfigColorView()
        view.title = "backgroundColor"
        view.color = .white
        view.colorChanged = { [weak self] in
            self?.config?.backgroundColor = $0
        }
        
        return view
    }()
    
    private lazy var bottomLineColorView: QDConfigColorView = {
        let view = QDConfigColorView()
        view.title = "bottomLineColor"
        view.color = .red
        
        view.colorChanged = { [weak self] in
            self?.config?.bottomLineColor = $0
        }
        
        
        return view
    }()
    
    private lazy var barHiddenView: QDConfigSwitchView = {
        let view = QDConfigSwitchView()
        view.title = "barHidden"
        view.on = true
        
        view.valueChanged = { [weak self] in
            self?.config?.barHidden = $0
        }
        
        return view
    }()
    
    private lazy var translucentView: QDConfigSwitchView = {
        let view = QDConfigSwitchView()
        view.title = "needBlurEffect"
        view.on = true
        view.valueChanged = { [weak self] in
            self?.config?.needBlurEffect = $0
        }
        return view
    }()
    
    private lazy var statusBarHiddenView: QDConfigSwitchView = {
        let view = QDConfigSwitchView()
        view.title = "statusBarHidden"
        view.on = true
        
        view.valueChanged = { [weak self] in
            self?.config?.statusBarHidden = $0
        }
        
        return view
    }()
    
    private lazy var statusBarStyleView: QDConfigSegmentView = {
        let items: [String]
        if #available(iOS 13.0, *) {
            items = ["Default", "Light", "Dark"]
        } else {
            items = ["Default", "Light"]
        }
        let view = QDConfigSegmentView(items: items)
        view.title = "statusBarStyle"
        view.selectedIndex = 0
                
        view.valueChanged = { [weak self] selectedIndex in
            let styles = [UIStatusBarStyle.default, .lightContent, .darkContent]
            let selectedStyle = styles[selectedIndex]
            self?.config?.statusBarStyle = selectedStyle
        }
        
        return view
    }()
    
    private lazy var bgImageView: QDConfigSegmentView = {
        let sunImage = UIImage(named: "sun_thumb")?.withRenderingMode(.alwaysOriginal)
        let rainImage = UIImage(named: "rain_thumb")?.withRenderingMode(.alwaysOriginal)
        let view = QDConfigSegmentView(items: ["none", sunImage as Any, rainImage as Any])
        view.title = "backgroundImage"
        view.selectedIndex = 0
        
        view.valueChanged = { [weak self] selectedIndex in
            let images: [UIImage?] = [nil, self?.sunBgImage, self?.rainBgImage]
            self?.config?.backgroundImage = images[selectedIndex]
        }
        return view
    }()
    
    private var sunBgImage: UIImage?
    private var rainBgImage: UIImage?
    
    var config: QDNavigationBarConfig? {
        didSet {
            updateView()
        }
    }
    
    // MARK: - Initialization
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        sunBgImage = UIImage(named: "sun_nav")
        rainBgImage = UIImage(named: "rain_nav")
        configSubviews()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        sunBgImage = UIImage(named: "sun_nav")
        rainBgImage = UIImage(named: "rain_nav")
        configSubviews()
    }
    
    // MARK: - Private Methods
    
    private func configSubviews() {
        addSubview(stackView)
        stackView.snp.makeConstraints { make in
            make.edges.equalToSuperview().inset(UIEdgeInsets(top: 8, left: 0, bottom: 5, right: 0))
        }
        
        let views: [UIView] = [
            bgColorView,
            bottomLineColorView,
            barHiddenView,
            translucentView,
            statusBarHiddenView,
            statusBarStyleView,
            bgImageView
        ]
        
        for view in views {
            stackView.addArrangedSubview(view)
            view.snp.makeConstraints { make in
                make.height.equalTo(48)
            }
        }
        
        layer.borderColor = UIColor.lightGray.cgColor
        layer.borderWidth = 1.0 / UIScreen.main.scale
        layer.cornerRadius = 15
    }
    
    private func updateView() {
        guard let config = config else { return }
        
        bgColorView.color = config.backgroundColor
        bottomLineColorView.color = config.bottomLineColor
        barHiddenView.on = config.barHidden
        translucentView.on = config.needBlurEffect
        statusBarHiddenView.on = config.statusBarHidden
        
        statusBarStyleView.selectedIndex = {
            switch config.statusBarStyle {
            case .default:
                return 0
            case .lightContent:
                return 1
            case .darkContent:
                return 2
            @unknown default:
                return 0
            }
        }()
        
        bgImageView.selectedIndex = {
            if config.backgroundImage == nil {
                return 0
            } else if config.backgroundImage == sunBgImage {
                return 1
            } else {
                return 2
            }
        }()
    }
    
    // MARK: - Getter & Setter
    
    private func setupColorView(_ view: QDConfigColorView, title: String, defaultColor: UIColor, colorChangedHandler: @escaping (UIColor) -> Void) -> QDConfigColorView {
        view.title = title
        view.color = defaultColor
        view.colorChanged = { color in
            colorChangedHandler(color)
        }
        return view
    }
    
    private func setupSwitchView(_ view: QDConfigSwitchView, title: String, defaultOn: Bool, valueChangedHandler: @escaping (Bool) -> Void) -> QDConfigSwitchView {
        view.title = title
        view.on = defaultOn
        view.valueChanged = { on in
            valueChangedHandler(on)
        }
        return view
    }
    
    private func setupSegmentView(_ view: QDConfigSegmentView, title: String, defaultIndex: Int, valueChangedHandler: @escaping (Int) -> Void) -> QDConfigSegmentView {
        view.title = title
        view.selectedIndex = defaultIndex
        view.valueChanged = { selectedIndex in
            valueChangedHandler(selectedIndex)
        }
        return view
    }
    
    // MARK: - View Lifecycle
    
    override func awakeFromNib() {
        super.awakeFromNib()
        configSubviews()
    }
}

